File: src/main/java/de/hanno/hpengine/util/multithreading/TripleBuffer.java
package de.hanno.hpengine.util.multithreading;

import de.hanno.hpengine.util.commandqueue.CommandQueue;

import java.util.concurrent.locks.ReentrantLock;
import java.util.function.Consumer;
import java.util.logging.Level;
import java.util.logging.Logger;

public class TripleBuffer<T extends TripleBuffer.State<T>> {
    private static final Logger LOGGER = Logger.getLogger(TripleBuffer.class.getName());

    private final QueueStatePair<T> instanceA;
    private final QueueStatePair<T> instanceB;
    private final QueueStatePair<T> instanceC;

    private ReentrantLock swapLock = new ReentrantLock();
    private ReentrantLock stagingLock = new ReentrantLock();

    private QueueStatePair<T> currentReadState;
    private QueueStatePair<T> currentWriteState;
    private QueueStatePair<T> currentStagingState;
    private QueueStatePair<T> temp;

    public TripleBuffer(T instanceA, T instanceB, T instanceC) {
        if(instanceA == null || instanceB == null || instanceC == null) {
            throw new IllegalArgumentException("Don't pass null to constructor!");
        }
        currentReadState = new QueueStatePair<>(instanceA);
        currentWriteState = new QueueStatePair<>(instanceB);
        currentStagingState = new QueueStatePair<>(instanceC);
        this.instanceA = currentReadState;
        this.instanceB = currentWriteState;
        this.instanceC = currentStagingState;
    }

    protected void swap() {
        swapLock.lock();
        stagingLock.lock();

        if(!currentReadState.state.shouldNotSwap(currentStagingState.state, currentReadState.state)) {
            temp = currentReadState;
            currentReadState = currentStagingState;
            currentStagingState = temp;
        }

        stagingLock.unlock();
        swapLock.unlock();
    }
    protected void swapStaging() {
        stagingLock.lock();

        temp = currentStagingState;
        currentStagingState = currentWriteState;
        currentWriteState = temp;

        stagingLock.unlock();
    }

    public void addCommand(Consumer<T> command) {
        instanceA.addCommand(command);
        instanceB.addCommand(command);
        instanceC.addCommand(command);
    }

    public boolean update() {
        currentWriteState.queue.executeCommands();
        swapStaging();
        return true;
    }

    public T getCurrentReadState() {
        return currentReadState.state;
    }
    public T getCurrentWriteState() {
        return currentWriteState.state;
    }

    public void startRead() {
        swapLock.lock();
        logState();
    }

    public void stopRead() {
        swap();
    }

    public void logState() {
        if(LOGGER.isLoggable(Level.FINER)) {
            LOGGER.fine("Read  " + (currentReadState == instanceA ? 0 : (currentReadState == instanceB ? 1 : 2)));
            LOGGER.fine("Stage " + (currentStagingState == instanceA ? 0 : (currentStagingState == instanceB ? 1 : 2)));
            LOGGER.fine("Write " + (currentWriteState == instanceA ? 0 : (currentWriteState == instanceB ? 1 : 2)));
        }
    }

    private static class QueueStatePair<T> {
        private final CommandQueue queue = new CommandQueue();
        private final T state;

        public QueueStatePair(T state) {
            this.state = state;
        }

        public void addCommand(Consumer<T> command) {
            queue.addCommand(() -> command.accept(state));
        }
    }

    public interface State<T> {
        boolean shouldNotSwap(T currentStaging, T currentRead);
    }
}

File: engine/src/main/kotlin/de/hanno/hpengine/graphics/state/multithreading/TripleBuffer.kt
package de.hanno.hpengine.graphics.state.multithreading

import de.hanno.hpengine.graphics.state.RenderState
import de.hanno.hpengine.graphics.state.StateRef
import java.util.concurrent.locks.ReentrantLock
import java.util.logging.Level
import java.util.logging.Logger
import kotlin.concurrent.withLock

class TripleBuffer constructor(
    private val instanceA: RenderState,
    private val instanceB: RenderState,
    private val instanceC: RenderState,
    private val preventSwap: (RenderState, RenderState) -> Boolean = { _, _ -> false }
) {

    constructor(
        factory: () -> RenderState,
        preventSwap: (RenderState, RenderState) -> Boolean = { _, _ -> false }
    ) : this(factory(), factory(), factory(), preventSwap)

    private val swapLock = ReentrantLock()
    private val stagingLock = ReentrantLock()

    var currentReadState: RenderState = instanceA
        private set
    var currentWriteState: RenderState = instanceB
        private set
    private var currentStagingState: RenderState = instanceC

    private var tempA: RenderState? = null
    private var tempB: RenderState? = null

    private fun swap(): Boolean = swapLock.withLock {
        stagingLock.withLock {
            if (!preventSwap(currentStagingState, currentReadState)) {
                tempA = currentReadState
                currentReadState = currentStagingState
                currentStagingState = tempA!!
                true
            } else false
        }
    }

    fun swapStaging() {
        stagingLock.withLock {
            tempB = currentStagingState
            currentStagingState = currentWriteState
            currentWriteState = tempB!!
        }
    }

    private var customStateCounter = 0
    fun <TYPE : Any> registerState(factory: () -> TYPE): StateRef<TYPE> {
        val newIndex = customStateCounter++
        instanceA.add(factory())
        instanceB.add(factory())
        instanceC.add(factory())
        return StateRef(newIndex)
    }

    fun readLocked(block: (RenderState) -> Unit) = swapLock.withLock {
        block(currentReadState)
        swap()
    }

    fun printState() {
        println("""Read  $currentReadStateIndex with cycle ${currentReadState.cycle}""")
        println("""Stage $currentStagingStateIndex with cycle ${currentStagingState.cycle}""")
        println("""Write $currentWriteStateIndex with cycle ${currentWriteState.cycle}""")
    }

    val currentReadStateIndex get() = if (currentReadState === instanceA) 0 else if (currentReadState === instanceB) 1 else 2
    val currentStagingStateIndex get() = if (currentStagingState === instanceA) 0 else if (currentStagingState === instanceB) 1 else 2
    val currentWriteStateIndex get() = if (currentWriteState === instanceA) 0 else if (currentWriteState === instanceB) 1 else 2

    val currentStateIndices get() = Triple(currentReadStateIndex, currentStagingStateIndex, currentWriteStateIndex)

}